app/app_main.c

```diff
@@ -2338,6 +2470,143 @@ static ACVP_RESULT app_rsa_handler(ACVP_TEST_CASE *test_case)
     return ACVP_SUCCESS;
 }
 
+/*
+ * RSA SigGen handler
+ */
+static ACVP_RESULT app_rsa_siggen_handler(ACVP_TEST_CASE *test_case)
+{
+    RSA *rsa = NULL;
+    EVP_MD *tc_md = NULL;
+    unsigned char *msg = NULL;
+    unsigned char *sigbuf = NULL;
+    int siglen, pad_mode;
+    long msglen = -1;
+    BIGNUM *bn_e;
+
+    ACVP_RSA_TC    *tc;
+
+    if (!test_case) {
+        return ACVP_INVALID_ARG;
+    }
+
+    tc = test_case->tc.rsa;
+
+    if (!tc) {
+        return ACVP_INVALID_ARG;
+    }
+
+    /*
+     * Set the message digest to the appropriate sha
+     */
+    if(strncmp(tc->sig_tc->sig_attrs_tc->hash_alg, ACVP_RSA_SHA_1, RSA_HASH_ALG_MAX_LEN ) == 0 ) {
+        tc_md = (EVP_MD *)EVP_sha1();
+    } else if(strncmp(tc->sig_tc->sig_attrs_tc->hash_alg, ACVP_RSA_SHA_224, RSA_HASH_ALG_MAX_LEN ) == 0 ) {
+        tc_md = (EVP_MD *)EVP_sha224();
+    } else if(strncmp(tc->sig_tc->sig_attrs_tc->hash_alg, ACVP_RSA_SHA_256, RSA_HASH_ALG_MAX_LEN ) == 0 ) {
+        tc_md = (EVP_MD *)EVP_sha256();
+    } else if(strncmp(tc->sig_tc->sig_attrs_tc->hash_alg, ACVP_RSA_SHA_384, RSA_HASH_ALG_MAX_LEN ) == 0 ) {
+        tc_md = (EVP_MD *)EVP_sha384();
+    } else if(strncmp(tc->sig_tc->sig_attrs_tc->hash_alg, ACVP_RSA_SHA_512, RSA_HASH_ALG_MAX_LEN ) == 0 ) {
+        tc_md = (EVP_MD *)EVP_sha512();
+    } else {
+        printf("\nError: hashAlg not supported for RSA SigGen\n");
+        return ACVP_INVALID_ARG;
+    }
+
+    /*
+     * Set the message given from the tc to binary form
+     */
+    msg = hex2bin_m(tc->sig_tc->sig_attrs_tc->msg, &msglen);
+    if (msglen == -1) {
+        printf("\nError: hex2bin error for RSA SigGen\n");
+        return ACVP_INVALID_ARG;
+    }
+
+    /*
+     * Make an RSA object and set a new BN exponent to use to generate a key
+     */
+    if (rsa) {
+        FIPS_rsa_free(rsa);
+    }
+
+    rsa = FIPS_rsa_new();
```

Comment:
Have we determined if this test is not available at runtime ? In general if you can perform it at runtime you don't use the FIPS calls, you would use the standard openssl call such as RSA_new() in this case.